### PR TITLE
bpo-30977: define uuid.UUID.__slots__ to save memory

### PR DESCRIPTION
**Pull request by wbolster**
https://bugs.python.org/issue30977

### CODE REVIEW

**Review by tiran — APPROVED** [review]
Looks sensible to me. What do you think, @warsaw ?

**Comment by amannijhawan** [comment]
Will this get backported to 2.7.x?

**Comment by wbolster** [comment]
hi, author here. is there anything i can do to help  move this pr forward?

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
This change breaks pickle. See my comments on the tracker.

File: Lib/uuid.py [hunk 1]
```diff
@@ -200,8 +202,8 @@ def __init__(self, hex=None, bytes=None, bytes_le=None, fields=None,
             # Set the version number.
             int &= ~(0xf000 << 64)
             int |= version << 76
-        self.__dict__['int'] = int
-        self.__dict__['is_safe'] = is_safe
+        super().__setattr__('int', int)
+        super().__setattr__('is_safe', is_safe)
```
Comment (vstinner) [inline Lib/uuid.py:206]: I suggest to use `object.__setattr__(self, 'int', int)`.

**Review by vstinner — APPROVED** [review]
LGTM, but please add a test in test_exceptions() to make sure that is_safe attribute cannot be modified, near the "badtype(lambda: setattr(u, 'int', i))" test.

**Comment by vstinner** [comment]
@serhiy-storchaka: "This change breaks pickle. See my comments on the tracker."

Oh, I missed this comment. It should be fixed.

Please rebase also your change.

**Comment by taleinat** [comment]
Since this has lingered for a long time with no word from the original author, I've gone ahead and implemented appropriate pickle/unpickle logic. I've created a new PR (#9078) since this one is old and has merge conflicts.

I suggest closing this.

**Comment by wbolster** [comment]
thanks all, especially @taleinat for finishing my first attempt!

**Pull request closed without merging**